In [1]:
import pandas as pd
import numpy as np

file_path = "../data/netflix_titles.csv"
df = pd.read_csv(file_path)

df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   show_id       8807 non-null   str  
 1   type          8807 non-null   str  
 2   title         8807 non-null   str  
 3   director      6173 non-null   str  
 4   cast          7982 non-null   str  
 5   country       7976 non-null   str  
 6   date_added    8797 non-null   str  
 7   release_year  8807 non-null   int64
 8   rating        8803 non-null   str  
 9   duration      8804 non-null   str  
 10  listed_in     8807 non-null   str  
 11  description   8807 non-null   str  
dtypes: int64(1), str(11)
memory usage: 825.8 KB


Why datatype for date_added is str instead of date
Some fields have null value

In [3]:
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


In [4]:
df.describe(include="object")

/var/folders/3p/13n6clgx5432cc1mbk90nkv00000gn/T/ipykernel_18713/702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,show_id,type,title,director,cast,country,date_added,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8803,8804,8807,8807
unique,8807,2,8807,4528,7692,748,1767,17,220,514,8775
top,s1,Movie,Dick Johnson Is Dead,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,1,19,19,2818,109,3207,1793,362,4


How come there is 748 unique values for "country"? total number of countries in the world is 195! look for:
- case sensitivity
- space
- mis spelling

In [5]:
# count missing values
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [6]:
# Calculate missing percentage
(df.isnull().sum()/len(df))*100

show_id          0.000000
type             0.000000
title            0.000000
director        29.908028
cast             9.367549
country          9.435676
date_added       0.113546
release_year     0.000000
rating           0.045418
duration         0.034064
listed_in        0.000000
description      0.000000
dtype: float64

Top 3 fields with the most missing data: director, then cast and country.

In [7]:
# any duolicates?
df.duplicated().sum()

np.int64(0)

## Summary of the Initial Dataset Observations

- Dataset contains 8,807 rows and 12 columns.
- `date_added` is stored as text and should be converted to a datetime type.
- The `director` column has a significant number of missing values.
- No duplicate rows were found.
- The dataset includes both numerical and categorical columns.
- The earliest release year is XXXX and the latest is YYYY.

# Phase 2: Cleaning Data

In [8]:
# Make a copy of the original data
df_clean = df.copy()

In [9]:
df_clean.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [ ]:
# Let'd decide about some of these missing values
# this dataset is mostly text so replacing missing values with a default text makes sense
# if the value was numerical we could .fillna(mean) or .fillna(median) based on the logic of situation
df_clean["director"] = df_clean["director"].fillna("Unknown")

In [12]:
df_clean.dtypes

show_id           str
type              str
title             str
director          str
cast              str
country           str
date_added        str
release_year    int64
rating            str
duration          str
listed_in         str
description       str
dtype: object

In [14]:
df_clean["date_added"] = pd.to_datetime(df_clean["date_added"], format="mixed")

In the above cell at first I didn't pass the format argument, it threw an error about the format of a few records that it couldn't match. with this argument I stated that there could be inconsistancies in different rows of this value and expected it to comprehend this. 
In real project I should keep record of these records and make sure the data/time is converted accordingly. As another approach I may use another format to enforce a specific conversion.

In [15]:
df_clean.dtypes

show_id                    str
type                       str
title                      str
director                   str
cast                       str
country                    str
date_added      datetime64[us]
release_year             int64
rating                     str
duration                   str
listed_in                  str
description                str
dtype: object

## Feature engineering
Let's add year, month, and day of adding the data from the date_added

In [19]:
df_clean["year_added"] = df_clean["date_added"].dt.year
df_clean["month_added"] = df_clean["date_added"].dt.month_name()
df_clean["day_added"] = df_clean["date_added"].dt.day_name()

In [20]:
df_clean.sample(10)

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added,month_added,day_added
2246,s2247,TV Show,Down to Earth with Zac Efron,Unknown,"Zac Efron, Darin Olien",United States,2020-07-10,2020,TV-PG,1 Season,"Docuseries, Reality TV, Science & Nature TV",Actor Zac Efron journeys around the world with...,2020.0,July,Friday
8307,s8308,Movie,The Forbidden Kingdom,Rob Minkoff,"Jackie Chan, Jet Li, Collin Chou, Liu Yifei, L...","United States, China",2019-07-01,2008,PG-13,104 min,"Action & Adventure, Sci-Fi & Fantasy",Jackie Chan and Jet Li star in this rousing ad...,2019.0,July,Monday
3168,s3169,TV Show,Fuller House,Unknown,"Candace Cameron Bure, Jodie Sweetin, Andrea Ba...",United States,2019-12-06,2020,TV-PG,5 Seasons,TV Comedies,The Tanner family’s adventures continue as DJ ...,2019.0,December,Friday
4423,s4424,Movie,ABCD: Any Body Can Dance,Remo D'Souza,"Prabhu Deva, Kay Kay Menon, Ganesh Acharya, Dh...",India,2018-11-01,2013,TV-14,136 min,"Dramas, International Movies",Thrown out of the elite dance studio he founde...,2018.0,November,Thursday
8278,s8279,Movie,The Devil We Know,"Stephanie Soechtig, Jeremy Seifert",NaN,United States,2019-01-15,2018,TV-14,88 min,Documentaries,This documentary details DuPont's alleged deca...,2019.0,January,Tuesday
3516,s3517,Movie,Oh! Baby,B. V. Nandini Reddy,"Samantha Ruth Prabhu, Lakshmi, Rajendraprasad,...",India,2019-09-14,2019,TV-14,157 min,"Comedies, International Movies, Music & Musicals",A surly septuagenarian gets another chance at ...,2019.0,September,Saturday
5013,s5014,Movie,Something Huge,Carlo Padial,"Berto Romero, Carolina Bang, Javier Botet, Car...",Spain,2018-03-01,2017,TV-MA,87 min,"Comedies, Independent Movies, International Mo...","A director and a comedian want to shoot the ""c...",2018.0,March,Thursday
2502,s2503,Movie,Escaping Tel Aviv,Sherif Arafa,"Karim Abdel Aziz, Mona Zaki, Sherif Mounir, En...",Egypt,2020-05-19,2009,TV-14,120 min,"Action & Adventure, International Movies",After her husband reveals he's an undercover M...,2020.0,May,Tuesday
6896,s6897,Movie,Green Room,Jeremy Saulnier,"Anton Yelchin, Imogen Poots, Patrick Stewart, ...",United States,2018-11-12,2015,R,95 min,"Horror Movies, Thrillers","At the end of their low-budget tour, a punk ba...",2018.0,November,Monday
4464,s4465,Movie,Welcome to Sajjanpur,Shyam Benegal,"Shreyas Talpade, Amrita Rao, Ravi Kishan, Ila ...",India,2018-11-01,2008,TV-14,131 min,"Comedies, Dramas, International Movies",An educated man assists the illiterate residen...,2018.0,November,Thursday


Now let's address the inconsistancies in the country.

In [ ]:
df_clean["country"].value_counts().head(50)
# so we see in addition to the reasons we named above causing to have too many distinct countries
# another reason is there is movies made by more than one country

country
United States                    2818
India                             972
United Kingdom                    419
Japan                             245
South Korea                       199
Canada                            181
Spain                             145
France                            124
Mexico                            110
Egypt                             106
Turkey                            105
Nigeria                            95
Australia                          87
Taiwan                             81
Indonesia                          79
Brazil                             77
United Kingdom, United States      75
Philippines                        75
United States, Canada              73
Germany                            67
China                              66
Thailand                           61
Argentina                          56
Hong Kong                          53
United States, United Kingdom      47
Italy                              45
Cana

In [25]:
#first let's clear hidden spaces
df_clean["country"] = df_clean["country"].str.strip()

In [27]:
df_clean["country"].nunique()

748

distinct values didn't change: 748, so space wasn't the problem

In [ ]:
# let's deal with case sensitivity
df_clean["country"] = df_clean["country"].str.lower()

In [29]:
df_clean["country"].nunique()

748

In [32]:
#let's do these for all columns
#find all text-based columns
string_cols = df_clean.select_dtypes(include=['object', 'string']).columns
df_clean[string_cols] = df_clean[string_cols].apply(lambda x: x.str.lower())
df_clean[string_cols] = df_clean[string_cols].apply(lambda x: x.str.strip())

In [49]:
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8807 non-null   str           
 1   type          8807 non-null   str           
 2   title         8807 non-null   str           
 3   director      8807 non-null   str           
 4   cast          7982 non-null   str           
 5   country       7976 non-null   str           
 6   date_added    8797 non-null   datetime64[us]
 7   release_year  8807 non-null   int64         
 8   rating        8803 non-null   str           
 9   duration      8804 non-null   str           
 10  genres        8807 non-null   str           
 11  description   8807 non-null   str           
 12  year_added    8797 non-null   float64       
 13  month_added   8797 non-null   str           
 14  day_added     8797 non-null   str           
dtypes: datetime64[us](1), float64(1), int64(1), str(1

In [37]:
df_clean.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'genres', 'description',
       'year_added', 'month_added', 'day_added'],
      dtype='str')

In [ ]:
#to increse readability of "listed_in". it makes more sense if it is named "genres"
df_clean.rename(
    columns={
        "listed_in": "genres"
    },
    inplace=True
)

In [ ]:
df_clean["release_year"].max()

np.int64(1925)

In [ ]:

df_clean["release_year"].min()

In [40]:
df_clean.to_csv(
    "../data/netflix_titles_clean.csv",
    index=False
)

In [41]:
df_check = pd.read_csv(
    "../data/netflix_titles_clean.csv"
)

df_check.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,genres,description,year_added,month_added,day_added
0,s1,movie,dick johnson is dead,kirsten johnson,NaN,united states,2021-09-25,2020,pg-13,90 min,documentaries,"as her father nears the end of his life, filmm...",2021.0,september,saturday
1,s2,tv show,blood & water,unknown,"ama qamata, khosi ngema, gail mabalane, thaban...",south africa,2021-09-24,2021,tv-ma,2 seasons,"international tv shows, tv dramas, tv mysteries","after crossing paths at a party, a cape town t...",2021.0,september,friday
2,s3,tv show,ganglands,julien leclercq,"sami bouajila, tracy gotoas, samuel jouy, nabi...",NaN,2021-09-24,2021,tv-ma,1 season,"crime tv shows, international tv shows, tv act...",to protect his family from a powerful drug lor...,2021.0,september,friday
3,s4,tv show,jailbirds new orleans,unknown,NaN,NaN,2021-09-24,2021,tv-ma,1 season,"docuseries, reality tv","feuds, flirtations and toilet talk go down amo...",2021.0,september,friday
4,s5,tv show,kota factory,unknown,"mayur more, jitendra kumar, ranjan raj, alam k...",india,2021-09-24,2021,tv-ma,2 seasons,"international tv shows, romantic tv shows, tv ...",in a city of coaching centers known to train i...,2021.0,september,friday


In [48]:
df_check.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   show_id       8807 non-null   str    
 1   type          8807 non-null   str    
 2   title         8807 non-null   str    
 3   director      8807 non-null   str    
 4   cast          7982 non-null   str    
 5   country       7976 non-null   str    
 6   date_added    8797 non-null   str    
 7   release_year  8807 non-null   int64  
 8   rating        8803 non-null   str    
 9   duration      8804 non-null   str    
 10  genres        8807 non-null   str    
 11  description   8807 non-null   str    
 12  year_added    8797 non-null   float64
 13  month_added   8797 non-null   str    
 14  day_added     8797 non-null   str    
dtypes: float64(1), int64(1), str(13)
memory usage: 1.0 MB


#Data cleaning summary
# Data Cleaning Summary

## Missing Values
- Filled missing directors with "Unknown"
- Kept missing cast values because they may be legitimate

## Duplicates
- There were no duplicates

## Data Types
- Converted date_added to datetime

## Text Cleaning
- Removed leading/trailing spaces
- All text columns became lower case

## Feature Engineering
- Created year_added
- Created month_added
- Created day_added


# Day 3: Exploratory Data Analysis


In [44]:
import pandas as pd
import numpy as np

In [45]:
df = pd.read_csv("../data/netflix_titles_clean.csv")

In [52]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   show_id       8807 non-null   str    
 1   type          8807 non-null   str    
 2   title         8807 non-null   str    
 3   director      8807 non-null   str    
 4   cast          7982 non-null   str    
 5   country       7976 non-null   str    
 6   date_added    8797 non-null   str    
 7   release_year  8807 non-null   int64  
 8   rating        8803 non-null   str    
 9   duration      8804 non-null   str    
 10  genres        8807 non-null   str    
 11  description   8807 non-null   str    
 12  year_added    8797 non-null   float64
 13  month_added   8797 non-null   str    
 14  day_added     8797 non-null   str    
dtypes: float64(1), int64(1), str(13)
memory usage: 1.0 MB


date_added has become string again, so:

In [53]:
df = pd.read_csv(
    "../data/netflix_titles_clean.csv",
    parse_dates=["date_added"]
)

In [54]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8807 non-null   str           
 1   type          8807 non-null   str           
 2   title         8807 non-null   str           
 3   director      8807 non-null   str           
 4   cast          7982 non-null   str           
 5   country       7976 non-null   str           
 6   date_added    8797 non-null   datetime64[us]
 7   release_year  8807 non-null   int64         
 8   rating        8803 non-null   str           
 9   duration      8804 non-null   str           
 10  genres        8807 non-null   str           
 11  description   8807 non-null   str           
 12  year_added    8797 non-null   float64       
 13  month_added   8797 non-null   str           
 14  day_added     8797 non-null   str           
dtypes: datetime64[us](1), float64(1), int64(1), str(1

In [56]:
df.shape

(8807, 15)

In [57]:
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'genres', 'description',
       'year_added', 'month_added', 'day_added'],
      dtype='str')

In [58]:
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,genres,description,year_added,month_added,day_added
0,s1,movie,dick johnson is dead,kirsten johnson,NaN,united states,2021-09-25,2020,pg-13,90 min,documentaries,"as her father nears the end of his life, filmm...",2021.0,september,saturday
1,s2,tv show,blood & water,unknown,"ama qamata, khosi ngema, gail mabalane, thaban...",south africa,2021-09-24,2021,tv-ma,2 seasons,"international tv shows, tv dramas, tv mysteries","after crossing paths at a party, a cape town t...",2021.0,september,friday
2,s3,tv show,ganglands,julien leclercq,"sami bouajila, tracy gotoas, samuel jouy, nabi...",NaN,2021-09-24,2021,tv-ma,1 season,"crime tv shows, international tv shows, tv act...",to protect his family from a powerful drug lor...,2021.0,september,friday
3,s4,tv show,jailbirds new orleans,unknown,NaN,NaN,2021-09-24,2021,tv-ma,1 season,"docuseries, reality tv","feuds, flirtations and toilet talk go down amo...",2021.0,september,friday
4,s5,tv show,kota factory,unknown,"mayur more, jitendra kumar, ranjan raj, alam k...",india,2021-09-24,2021,tv-ma,2 seasons,"international tv shows, romantic tv shows, tv ...",in a city of coaching centers known to train i...,2021.0,september,friday


# Q1: What type of content does Netflix have more of?

In [59]:
df["type"].value_counts()

type
movie      6131
tv show    2676
Name: count, dtype: int64

# Q2: By roughly what percentage?

In [62]:
df["type"].value_counts(normalize=True) * 100

type
movie      69.615079
tv show    30.384921
Name: proportion, dtype: float64

# Q3: Explore release year

In [63]:
df["release_year"].describe()

count    8807.000000
mean     2014.180198
std         8.819312
min      1925.000000
25%      2013.000000
50%      2017.000000
75%      2019.000000
max      2021.000000
Name: release_year, dtype: float64

In [64]:
df.nsmallest(5, "release_year")

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,genres,description,year_added,month_added,day_added
4250,s4251,tv show,pioneers: first women filmmakers*,unknown,NaN,NaN,2018-12-30,1925,tv-14,1 season,tv shows,this collection restores films from women who ...,2018.0,december,sunday
7790,s7791,movie,prelude to war,frank capra,NaN,united states,2017-03-31,1942,tv-14,52 min,"classic movies, documentaries",frank capra's documentary chronicles the rise ...,2017.0,march,friday
8205,s8206,movie,the battle of midway,john ford,"henry fonda, jane darwell",united states,2017-03-31,1942,tv-14,18 min,"classic movies, documentaries",director john ford captures combat footage of ...,2017.0,march,friday
8660,s8661,movie,undercover: how to operate behind enemy lines,john ford,NaN,united states,2017-03-31,1943,tv-pg,61 min,"classic movies, documentaries",this world war ii-era training film dramatizes...,2017.0,march,friday
8739,s8740,movie,why we fight: the battle of russia,"frank capra, anatole litvak",NaN,united states,2017-03-31,1943,tv-pg,82 min,documentaries,this installment of frank capra's acclaimed do...,2017.0,march,friday


In [65]:
df.nlargest(5, "release_year")

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,genres,description,year_added,month_added,day_added
1,s2,tv show,blood & water,unknown,"ama qamata, khosi ngema, gail mabalane, thaban...",south africa,2021-09-24,2021,tv-ma,2 seasons,"international tv shows, tv dramas, tv mysteries","after crossing paths at a party, a cape town t...",2021.0,september,friday
2,s3,tv show,ganglands,julien leclercq,"sami bouajila, tracy gotoas, samuel jouy, nabi...",NaN,2021-09-24,2021,tv-ma,1 season,"crime tv shows, international tv shows, tv act...",to protect his family from a powerful drug lor...,2021.0,september,friday
3,s4,tv show,jailbirds new orleans,unknown,NaN,NaN,2021-09-24,2021,tv-ma,1 season,"docuseries, reality tv","feuds, flirtations and toilet talk go down amo...",2021.0,september,friday
4,s5,tv show,kota factory,unknown,"mayur more, jitendra kumar, ranjan raj, alam k...",india,2021-09-24,2021,tv-ma,2 seasons,"international tv shows, romantic tv shows, tv ...",in a city of coaching centers known to train i...,2021.0,september,friday
5,s6,tv show,midnight mass,mike flanagan,"kate siegel, zach gilford, hamish linklater, h...",NaN,2021-09-24,2021,tv-ma,1 season,"tv dramas, tv horror, tv mysteries",the arrival of a charismatic young priest brin...,2021.0,september,friday


# Q4: Which audience rating appears most often?

In [66]:
df["rating"].value_counts()

rating
tv-ma       3207
tv-14       2160
tv-pg        863
r            799
pg-13        490
tv-y7        334
tv-y         307
pg           287
tv-g         220
nr            80
g             41
tv-y7-fv       6
nc-17          3
ur             3
74 min         1
84 min         1
66 min         1
Name: count, dtype: int64

In [67]:
df["rating"].value_counts(normalize=True) * 100

rating
tv-ma       36.430762
tv-14       24.537090
tv-pg        9.803476
r            9.076451
pg-13        5.566284
tv-y7        3.794161
tv-y         3.487447
pg           3.260252
tv-g         2.499148
nr           0.908781
g            0.465750
tv-y7-fv     0.068159
nc-17        0.034079
ur           0.034079
74 min       0.011360
84 min       0.011360
66 min       0.011360
Name: proportion, dtype: float64

# Analyze Countries
# Q5: Which countries produce the most Netflix content?

In [68]:
(df["country"].value_counts(normalize=True) * 100).head(10)

country
united states     35.330993
india             12.186560
united kingdom     5.253260
japan              3.071715
south korea        2.494985
canada             2.269308
spain              1.817954
france             1.554664
mexico             1.379137
egypt              1.328987
Name: proportion, dtype: float64

In [69]:
df["country"].value_counts().head(10)

country
united states     2818
india              972
united kingdom     419
japan              245
south korea        199
canada             181
spain              145
france             124
mexico             110
egypt              106
Name: count, dtype: int64

In [72]:
print((df['country'] == 'united states, canada').sum())

73


# Review aggregation function

In [73]:
df.groupby("type")["release_year"].agg(
    ["count", "min", "max", "mean"]
)

,count,min,max,mean
type,,,,
movie,6131,1942,2021,2013.121514
tv show,2676,1925,2021,2016.605755
